scipio/src/channels/local_channel.rs

```diff
@@ -0,0 +1,586 @@
+    ///     sender.send(0);
+    ///     assert_eq!(sender.has_room(), false);
+    ///     receiver.next().await.unwrap();
+    ///     assert_eq!(sender.has_room(), true);
+    /// });
+    /// ```
+    ///
+    /// [`LocalChannel`]: struct.LocalChannel.html
+    pub fn new_bounded(size: usize) -> (LocalSender<T>, LocalReceiver<T>) {
+        Self::new(ChannelCapacity::Bounded(size))
+    }
+}
+
+struct SendWaiter<T> {
+    channel: LocalChannel<T>,
+}
+
+impl<T> Future for SendWaiter<T> {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.channel.has_room() {
+            Poll::Ready(())
+        } else {
+            let mut state = self.channel.state.borrow_mut();
+            state.send_waiters.push(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+}
+
+impl<T> LocalSender<T> {
+    /// Sends data into this channel.
+    ///
+    /// It returns [`io::Error`] [`BrokenPipe`] if the receiver is destroyed.
+    /// It returns [`io::Error`] [`Other`] if this is a bounded channel that has no more capacity
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::{LocalExecutor, Local};
+    /// use scipio::channels::LocalChannel;
+    /// use futures::stream::StreamExt;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    /// ex.run(async move {
+    ///     let (sender, mut receiver) = LocalChannel::new_bounded(1);
+    ///     sender.send(0);
+    ///     sender.send(0).unwrap_err(); // no more capacity
+    ///     receiver.next().await.unwrap(); // now we have capacity again
+    ///     drop(receiver); // but because the receiver is destroyed send will err
+    ///     sender.send(0).unwrap_err();
+    /// });
+    /// ```
+    ///
+    /// [`BrokenPipe`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.BrokenPipe
+    /// [`Other`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.Other
+    /// [`io::Error`]: https://doc.rust-lang.org/std/io/struct.Error.html
+    pub fn send(&self, item: T) -> io::Result<()> {
+        self.channel.push(item)
+    }
+
+    /// Sends data into this channel when it is ready to receive it
+    ///
+    /// For an unbounded channel this is just a more expensive version of [`send`]. Prefer
+    /// to use [`send`] instead.
+    ///
+    /// For a bounded channel this will push to the channel when the channel is ready to
+    /// receive data.
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::{LocalExecutor, Local};
+    /// use scipio::channels::LocalChannel;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    /// ex.run(async move {
+    ///     let (sender, receiver) = LocalChannel::new_bounded(1);
+    ///     sender.send_eventually(0).await.unwrap();
+    ///     drop(receiver);
+    /// });
+    /// ```
+    ///
+    /// [`send`]: struct.LocalSender.html#method.send
+    pub async fn send_eventually(&self, item: T) -> io::Result<()> {
```

Comment:
Idiomatic names here are send for blocking send and try_send for non-blocking send. I think the two best crates to study channel’s APIs are flume and crossbeam channel.

Reply:
Additional point of interest: there was a long discussion about the result of send. There are a regiments for the position that send should never fail, and that sending over a closed channel should be an error: https://github.com/crossbeam-rs/crossbeam/issues/314

Reply:
Thanks. In the case of an unbounded channel `try_send` would always succeed but I guess that's okay.